clients/cobol-lsp-vscode-extension/src/services/ProcessorGroups.ts

```diff
@@ -206,50 +220,107 @@ function pathMatches(program: string, documentPath: string) {
+    return undefined;
+  }
+  return fs.readFileSync(procCfgPath).toString();
+}
+
+export const readProgramConfigFileContent = (): string | undefined => {
+  const ws = SettingsUtils.getWorkspaceFoldersPath(true);
+  if (ws.length < 1) {
+    return undefined;
+  }
+  const cfgPath = path.join(ws[0], PROCESSOR_GROUP_FOLDER);
   const pgmCfgPath = path.join(cfgPath, PROCESSOR_GROUP_PGM);
-  if (!fs.existsSync(pgmCfgPath) || !fs.existsSync(procCfgPath)) {
+  if (!fs.existsSync(pgmCfgPath)) {
     return undefined;
   }
-  const procCfg: ProcessorsConfig = JSON.parse(
-    fs.readFileSync(procCfgPath).toString(),
-  );
-  const pgmCfg: ProgramsConfig = JSON.parse(
-    fs.readFileSync(pgmCfgPath).toString(),
-  );
-  const pgroup = matchProcessorGroup(pgmCfg, documentPath, ws[0]);
+  return fs.readFileSync(pgmCfgPath).toString();
+};
+
+export const loadProcessorsConfigForDocument = (
+  documentUri: string,
+  processorGroupsContent: string | undefined = readProcessorGroupsFileContent(),
+  programConfigContent: string | undefined = readProgramConfigFileContent(),
+  b4gContent: string | undefined = loadBridgeJsonContent(
+    Uri.parse(documentUri).fsPath,
+  ),
+): ProcessorConfig | undefined => {
+  const documentPath = Uri.parse(documentUri).fsPath;
+  const ws = SettingsUtils.getWorkspaceFoldersPath(true);
+  if (ws.length < 1) {
+    return undefined;
+  }
+  if (processorGroupsContent === undefined) {
+    return undefined;
+  }
+  const pgroups = loadProcessorConfigurations(processorGroupsContent);
+  const groupsConfig: ProcessorGroupConfiguration = {
+    groupMap: {},
+    procGroups: pgroups,
+  };
+  if (programConfigContent) {
+    const pgmCfg: ProgramsConfig = JSON.parse(programConfigContent);
+    pgmCfg.pgms.forEach((p) => {
+      groupsConfig.groupMap[p.program] = p.pgroup;
+    });
+  }
+
+  const b4g = b4gContent ? decodeBridgeJson(b4gContent) : undefined;
+  const pgroup = selectProcessorGroup(groupsConfig, documentPath, ws[0], b4g);
 
   let result;
-  procCfg.pgroups.forEach((p) => {
+  pgroups.forEach((p) => {
     if (pgroup === p.name) {
       result = p;
       return;
     }
   });
   return result;
+};
+
+function selectProcessorGroup(
+  pgmCfg: ProcessorGroupConfiguration,
+  documentPath: string,
+  workspacePath: string,
+  b4g: B4GTypeMetadata | undefined,
+): string | undefined {
+  if (b4g === undefined) {
+    return matchProcessorGroup(pgmCfg, documentPath, workspacePath);
+  }
+  const selectedElement = b4g.fileExtension
+    ? path.basename(documentPath, "." + b4g.fileExtension)
+    : path.basename(documentPath);
+  return b4g.elements[selectedElement] === undefined
+    ? b4g.defaultProcessorGroup
+    : b4g.elements[selectedElement].processorGroup;
 }
 
 function loadProcessorGroupSettings<T>(
-  scopeUri: string,
+  pgCfg: ProcessorConfig | undefined,
   atrtibute: string,
   configObject: T,
   dialect: string = "COBOL",
 ): T {
+  if (pgCfg === undefined) {
+    return configObject;
+  }
```

Comment:
All calls to `loadProcessorsConfigForDocument` but one are done to provide the first argument to this function...
```suggestion
  scopeUri: string,
  atrtibute: string,
  configObject: T,
  dialect: string = "COBOL",
): T {
  const pgCfg = loadProcessorsConfigForDocument(
    scopeUri,
    readProcessorGroupsFileContent(),
    readProgramConfigFileContent(),
    loadBridgeJsonContent(Uri.parse(documentUri).fsPath),
  );
  if (pgCfg === undefined) {
    return configObject;
  }
```